[US-2652] Different approach to multidomain entities

MattCzerner opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| Review of designed solution
|New feature| No
|BC breaks| Yes 
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

MattCzerner commented:
- BrandData now has responsibility for SEO properties
- BrandFormType now has reponsibility of SEOP properties
- BrandEditFormType deleted
- BrandEditData* deleted because it is useless now
- BrandDOmainFactory deleted since you can access brandDomain through entity or BrandData

PetrHeinz commented on composer.json:
```diff
@@ -111,7 +111,7 @@
     "presta/sitemap-bundle": "^1.5.3",
     "prezent/doctrine-translatable": "^1.2",
-    "prezent/doctrine-translatable-bundle": "^1.0.3",
+    "prezent/doctrine-translatable-bundle": "1.0.3",
```
What was the error and in which version of prezent/doctrine-translatable-bundle? You should definitely elaborate in the commit message.

TomasLudvik replied:
You do not need to use this anymore. This problem has been already fixed. Please remove this commit.

MattCzerner replied:
prezent released new version with fix of that error, ill drop this commit.

PetrHeinz commented on packages/framework/src/DataFixtures/Demo/BrandDataFixture.php:
```diff
@@ -65,6 +65,14 @@ public function load(ObjectManager $manager)
+            $brandData->seoMetaDescriptions[2] = null;
+            $brandData->seoH1s[1] = null;
+            $brandData->seoH1s[2] = null;
+
```
Demo data fixtures are executed before second domain is created, you should add the nulls only for Domain::FIRST_DOMAIN_ID and add the second domain in DemoMultidomain (or let it be automatically created during create-demo-data phing target)

MattCzerner replied:
I created DemoDomainFixture. I think that it belongs into create-demo-data but there is plan which will change behaviour of this target, so I will leave it to implementation of that US

PetrHeinz commented on packages/framework/src/Form/Admin/Product/Brand/BrandFormType.php:
```diff
@@ -3,24 +3,84 @@
     /**
      * @param \Symfony\Component\Form\FormBuilderInterface $builder
      * @param array $options
+     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
```
Don't forget to remove these annotations when https://github.com/shopsys/shopsys/pull/143 is merged. :)

MattCzerner replied:
I will delete this with rebase on current master. I dont want my build to fail at this time.

PetrHeinz commented on packages/framework/src/Model/Product/Brand/BrandDataFactory.php:
```diff
@@ -51,9 +65,13 @@ public function createFromBrand(Brand $brand)
         }
 
-        foreach ($brandDomains as $brandDomain) {
+        foreach ($brand->getDomains() as $brandDomain) {
```
Is it necessary to make the getter public? It might be better to keep the concept of BrandDomains inside the Brand itself. You can iterate over $this->domain->getAll(). See that you access the $brandDomain here only in order to get its domain ID.

MattCzerner replied:
I agree, now are all domain ids looped in this foreach

PetrHeinz commented on packages/framework/src/Model/Product/Brand/BrandFacade.php:
```diff
@@ -110,19 +99,16 @@ public function create(BrandEditData $brandEditData)
-
-        $this->refreshBrandDomains($brand, $brandEditData);
-
+        $this->em->persist($brand);
```
Entity cannot be responsible for persisting its secondary entities, because it doesn't have access to EntityManager. Entity is rather responsible just for creating the secondary entities, facade is responsible for persisting the created/updated entity so it starts the persistence cascade.

It's probably just semantics, I just want to be on the same page :)

PetrHeinz commented on packages/framework/src/Model/Transport/Transport.php:
```diff
@@ -244,6 +254,44 @@ public function setPosition($position)
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Transport\TransportData $transportData
+     */
+    private function setDomains(TransportData $transportData)
```
There should be no private methods in entities.

PetrHeinz commented on packages/framework/src/Model/Transport/Transport.php:
```diff
@@ -244,6 +254,44 @@ public function setPosition($position)
+            } catch (TransportDomainNotFoundException $e) {
+                $paymentDomain = new TransportDomain($this, $domainId);
+            }
+            $this->domains[$domainId] = $paymentDomain;
```
TransportDomain is not an entity for multidomain attributes (similar to ProductDomain), but a optional association with a domain (similar to PaymentDomain). This means, that you should be able to remove the association to a domain (when trying to switch the transport from the first domain to the second, for example), but the setDomains method doesn't allow it, it only adds new Transport domains.

TomasLudvik commented on packages/framework/src/Model/Product/ProductRepository.php:
```diff
@@ -165,10 +157,12 @@ public function addTranslation(QueryBuilder $queryBuilder, $locale)
+        $queryBuilder->addSelect('pd')
+            ->join('p.domains', 'pd', Join::WITH, 'pd.domainId = :domain');
+
+        $queryBuilder->setParameter('domain', $domainId);
```
I do not understand changes in this method

MattCzerner replied:
ProductRepository does some visibility calculations in which they control ProdudctDomain properties, thats why it is needed to be added. It is same concept as it is used for Translations.

TomasLudvik commented on packages/framework/src/Resources/views/Admin/Content/Product/visibilityItems.html.twig:
```diff
@@ -1,12 +1,13 @@
 <ul class="list-domain">
-    {% for productDomain in productDetail.productDomainsIndexedByDomainId %}
+    {% for product in productDetail.product %}
```
Are you now iterating products instead of domains?

TomasLudvik commented on packages/framework/src/DataFixtures/DemoMultidomain/CategoryRootDataFixture.php:
```diff
@@ -0,0 +1,50 @@
+<?php
```
There is no description why is this class added

TomasLudvik replied:
Like in CM

MattCzerner replied:
I will add it at the final rebase

TomasLudvik commented on project-base/tests/ShopBundle/Database/Model/Order/OrderTransportAndPaymentTest.php:
```diff
@@ -86,7 +83,7 @@ public function testVisibleTransportHiddenPayment()
         $transport = new Transport(new TransportData(['cs' => 'transportName', 'en' => 'transportName'], $vat, [], [], false));
         $transportDomain = new TransportDomain($transport, 1);
         $payment = new Payment(new PaymentData(['cs' => 'paymentName', 'en' => 'paymentName'], $vat, [], [], true));
```
You should set domain here

MattCzerner replied:
I refactored it entirely.

PetrHeinz commented on packages/framework/src/Model/Product/Brand/Brand.php:
```diff
@@ -91,6 +102,75 @@ protected function createTranslation()
+     */
+    protected function getBrandDomain(int $domainId)
+    {
+        foreach ($this->domains as $domain) {
```
The ProductDomains are not necessarily indexed by domain ID (and it's probably unwise to index it in such a way in https://github.com/shopsys/shopsys/pull/165/files/0ff35b922475cab1d02d6b2ce9a3b452d4906834#diff-ce1e32a3065400b38b9b2cac44e743e3R116). That's because Doctrine will hydrate the property as a unordered collection - it won't even keep the order of the array, let alone the indexes. Indexes are not persisted in any way by the Doctrine.

PetrHeinz commented on packages/framework/src/Model/Product/Product.php:
```diff
@@ -893,6 +905,109 @@ protected function setTranslations(ProductData $productData)
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\ProductData $productData
+     */
+    private function setDomains($productData)
```
Facades should have protected methods as they are supposed to be extended by inheritance.

PetrHeinz replied:
Also, the parameter should be type-hinted.

MattCzerner replied:
It is not facade, but i think that this rule applies to entities too. Fixed

PetrHeinz commented on packages/framework/src/Model/Product/Product.php:
```diff
@@ -893,6 +905,109 @@ protected function setTranslations(ProductData $productData)
+     * @param \Shopsys\FrameworkBundle\Component\Domain\Domain $domain
+     * @return string
+     */
+    public function getSeoTitleForHtml(Domain $domain)
```
Should this logic be in entity? In Brand, logic such as this was in the template.

MattCzerner replied:
Okay, moved to template and refactored the old implementation of getting seo title in brand since it was wierdly conditioned and it was incosistent with other implementations.

PetrHeinz commented on packages/framework/src/Model/Product/Product.php:
```diff
@@ -893,6 +905,109 @@ protected function setTranslations(ProductData $productData)
+     * @param int $domainId
+     * @return string|null
+     */
+    public function getSeoMetaDescription($domainId)
```
not type-hinted

PetrHeinz commented on packages/framework/src/Model/Product/Product.php:
```diff
@@ -893,6 +905,109 @@ protected function setTranslations(ProductData $productData)
+    /**
+     * @return \Shopsys\FrameworkBundle\Model\Product\ProductDomain[]|\Doctrine\Common\Collections\ArrayCollection
+     */
+    public function getDomains()
```
In Brand, method such as this wasn't public. The idea was that BrandDomain should look like an implementation detail from the outside.

MattCzerner replied:
But, Brand has this method public, you need to have this public because of `ProductDataFactory` that needs these domains in `createFromBrand()` method, I thought of it and i dont see any other solution for not requiring these Domain entities

PetrHeinz replied:
How about iterating over `Domain::getAllIds()` in the factory and accessing then accessing the attributes via the main entity - eg. `Product::getSeoTitle($domainId)`?

PetrHeinz commented on packages/framework/src/Model/Product/Product.php:
```diff
@@ -893,6 +905,109 @@ protected function setTranslations(ProductData $productData)
+     * @param int $domainId
+     * @return \Shopsys\FrameworkBundle\Model\Product\ProductDomain
+     */
+    public function getProductDomain(int $domainId)
```
In Brand, method such as this wasn't public. The idea was that BrandDomain should look like an implementation detail from the outside.

PetrHeinz commented on packages/framework/src/Model/Product/ProductDataFactory.php:
```diff
@@ -124,12 +124,23 @@ public function createFromProduct(Product $product)
             $variantAliases[$translation->getLocale()] = $translation->getVariantAlias();
         }
+
+        foreach ($domains as $domain) {
```
We usually call instances of DomainConfig a $domain, these objects should be called $productDomains imo to avoid confusion.

PetrHeinz commented on packages/framework/src/Model/Product/ProductRepository.php:
```diff
@@ -162,10 +154,12 @@ public function addTranslation(QueryBuilder $queryBuilder, $locale)
      * @param int $domainId
      */
-    public function addDomain(QueryBuilder $queryBuilder, $domainId)
+    private function addDomain(QueryBuilder $queryBuilder, $domainId)
```
Generally, there shouldn't be private methods in repositories as they are supposed to be extended by inheritance.

PetrHeinz commented on packages/framework/src/Model/Product/ProductRepository.php:
```diff
@@ -162,10 +154,12 @@ public function addTranslation(QueryBuilder $queryBuilder, $locale)
-        $queryBuilder->join(ProductDomain::class, 'pd', Join::WITH, 'pd.product = p AND pd.domainId = :domainId');
-        $queryBuilder->setParameter('domainId', $domainId);
+        $queryBuilder->addSelect('pd')
+            ->join('p.domains', 'pd', Join::WITH, 'pd.domainId = :domain');
```
Why did you rename the parameter `domainId` to `domain`?

MattCzerner replied:
ups, fixed

PetrHeinz commented on packages/framework/src/DataFixtures/Base/CategoryRootDataFixture.php:
```diff
@@ -20,36 +19,35 @@ class CategoryRootDataFixture extends AbstractReferenceFixture implements Depend
-        $rootCategory = $this->categoryFactory->create(new CategoryData());
+        $categoryData = $this->categoryDataFactory->createDefault();
+        $categoryData->enabled = [
+            Domain::FIRST_DOMAIN_ID => true,
```
Shouldn't this behavior be a part of CategoryDataFactory::createDefault()?

MattCzerner replied:
well not really, `createDefault()` creates categoryData with enabled set to false. So this line makes sure that category will be enabled on domain

PetrHeinz replied:
Why is it created with `false`s as default? Isn't more appropriate to create it with `true`s? On `master`, all domains are enabled by default when eg. creating a new category:

![image](https://user-images.githubusercontent.com/10008612/41303387-685a3e20-6e6d-11e8-921b-0124a121f5d3.png)

MattCzerner replied:
Okay, i think that i handled all of our enabled domain properties and it is set to be enabled as default, which made data fixtures much more easier.

PetrHeinz commented on packages/framework/src/DataFixtures/DemoMultidomain/CategoryDataFixture.php:
```diff
@@ -128,6 +128,7 @@ private function editCategoryOnDomain2($referenceName, $descriptionDomain2)
         $categoryData = $this->categoryDataFacade->createFromCategory($category);
         $domainId = 2;
         $categoryData->descriptions[$domainId] = $descriptionDomain2;
+        $categoryData->enabled[$domainId] = true;
```
Is this necessary? The create-domain-data should take care of non-nullable multidomain parameters automatically.

MattCzerner replied:
Not now.

PetrHeinz commented on packages/framework/src/Migrations/Version20180611215810.php:
```diff
@@ -0,0 +1,24 @@
+     */
+    public function up(Schema $schema)
+    {
+        $this->sql('ALTER TABLE category_domains RENAME COLUMN hidden TO enabled');
```
This looks like it negates all current category visibility.

MattCzerner replied:
Good point I added UPDATE migration that will solve this problem.

PetrHeinz commented on packages/framework/src/Model/Category/Category.php:
```diff
@@ -210,11 +214,114 @@ protected function setTranslations(CategoryData $categoryData)
+     * @param int $domainId
+     * @return string|null
+     */
+    public function getSeoTitle($domainId)
```
In previous entities these methods were always type-hinted as `int`.

MattCzerner replied:
added type hint

PetrHeinz commented on packages/framework/src/Model/Category/Category.php:
```diff
@@ -210,11 +214,114 @@ protected function setTranslations(CategoryData $categoryData)
+    /**
+     * @return \Doctrine\Common\Collections\ArrayCollection|\Shopsys\FrameworkBundle\Model\Category\CategoryDomain[]
+     */
+    public function getCategoryDomains()
```
In Brand, this method wasn't public.

MattCzerner replied:
it was. as i said, i think that there is not other solution to make this method private

PetrHeinz commented on packages/framework/src/Model/Category/Category.php:
```diff
@@ -210,11 +214,114 @@ protected function setTranslations(CategoryData $categoryData)
+     * @param \Shopsys\FrameworkBundle\Model\Category\CategoryData $categoryData
+     * @return array
+     */
+    private function getDomainIdsForCategoryDomainData(CategoryData $categoryData)
```
Why not use just the `array_keys($categoryData->seoTitles)` as in Brand and Product? Provided data object should be valid - should have attributes set for every domain ID (for which the data object factory is responsible).

PetrHeinz commented on packages/framework/src/Component/Transformers/ShowOnDomainBooleanToIntegerTransformer.php:
```diff
@@ -0,0 +1,51 @@
+use Shopsys\FrameworkBundle\Component\Domain\Domain;
+use Symfony\Component\Form\DataTransformerInterface;
+
+class ShowOnDomainBooleanToIntegerTransformer implements DataTransformerInterface
```
This logic should be probably implemented right in the `DomainsType` - it is used in the three places only (Category, Payment, Transport). If it was just a collection of `CheckboxTypes`, the data would be in correct format without transformers.

PetrHeinz replied:
The DomainsType can be as simple as this, without the need of a transformer or any changes to current form types or theme:

```php
<?php

namespace Shopsys\FrameworkBundle\Form;

use Shopsys\FrameworkBundle\Component\Domain\Domain;
use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
use Symfony\Component\Form\FormBuilderInterface;

class DomainsType extends AbstractType
{
    /**
     * @var \Shopsys\FrameworkBundle\Component\Domain\Domain
     */
    private $domain;

    /**
     * @param \Shopsys\FrameworkBundle\Component\Domain\Domain $domain
     */
    public function __construct(Domain $domain)
    {
        $this->domain = $domain;
    }

    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        foreach ($this->domain->getAll() as $domainConfig) {
            $builder->add($domainConfig->getId(), CheckboxType::class, [
                'required' => false,
                'label' => $domainConfig->getName(),
            ]);
        }
    }
}

MattCzerner replied:
great

MattCzerner commented on packages/framework/src/Model/Product/Brand/BrandDataFactory.php:
```diff
@@ -51,8 +65,11 @@ public function createFromBrand(Brand $brand)
+        foreach ($allDomainIds as $domainId) {
+            $brandData->seoH1s[$domainId] = $brand->getSeoH1($domainId);
+            $brandData->seoTitles[$domainId] = $brand->getSeoTitle($domainId);
+            $brandData->seoMetaDescriptions[$domainId] = $brand->getSeoMetaDescription($domainId);
```
I dont understand on which line they got overwrited?

PetrHeinz replied:
https://github.com/shopsys/shopsys/blob/97583067168a4a4754f1a2a0c0a528732b8678a7/packages/framework/src/Model/Product/Brand/BrandDataFactory.php#L80-L82

PetrHeinz replied:
The overwrite was then removed in a15ef6c97e9440b597bc1cc7cc0f54387feaaf03, see history of the file: https://github.com/shopsys/shopsys/commits/mc-multidomain/packages/framework/src/Model/Product/Brand/BrandDataFactory.php

MattCzerner commented on packages/framework/src/Model/Product/Brand/Exception/BrandDomainNotFoundException.php:
```diff
@@ -0,0 +1,9 @@
+
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+class BrandDomainNotFoundException extends NotFoundHttpException implements BrandException
```
Okay, your right, now i throw 500 error and i added message into exception

MattCzerner commented on packages/framework/src/Form/Admin/Product/Brand/BrandFormType.php:
```diff
@@ -3,24 +3,83 @@
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $brand = $options['brand'];
```
You are right, changed

MattCzerner commented on packages/framework/src/DataFixtures/DemoMultidomain/BrandDataFixture.php:
```diff
@@ -0,0 +1,115 @@
+
+            $brandData->seoTitles[self::DOMAIN_ID] = null;
+            $brandData->seoMetaDescriptions[self::DOMAIN_ID] = null;
+            $brandData->seoH1s[self::DOMAIN_ID] = null;
```
Well you are right that this is actually useless and its already done by BrandDataFactory. But i dont think that this should be used for translating the name, i think that Brand name is not translatable.

PetrHeinz replied:
So there is no need for this data fixture?

MattCzerner replied:
Well, the fix of this is done via create-multidomain-data. in which i manually added all of our new domain entities, i know that this is not the best solution but i think that this is actually responsibility of other US.

PetrHeinz replied:
Please, add a note about these changes to *US-4163* so it's not forgotten.

MattCzerner commented on packages/framework/src/DataFixtures/Demo/BrandDataFixture.php:
```diff
@@ -65,6 +65,11 @@ public function load(ObjectManager $manager)
+
+            $brandData->seoTitles[1] = null;
+            $brandData->seoMetaDescriptions[1] = null;
+            $brandData->seoH1s[1] = null;
```
These 3 lines are actually useless, because it is done by BrandDataFactory.

PetrHeinz replied:
Yeah, deleting it would be much better :+1:

MattCzerner commented on packages/framework/src/DataFixtures/DemoMultidomain/BrandDataFixture.php:
```diff
@@ -0,0 +1,115 @@
+    const BRAND_ORAVA = 'brand_orava';
+    const BRAND_OLYMPUS = 'brand_olympus';
+    const BRAND_HYUNDAI = 'brand_hyundai';
+    const BRAND_NIKON = 'brand_nikon';
```
There is no reason, i just didnt do it. Fixed!

MattCzerner commented on packages/framework/src/Migrations/Version20180611123750.php:
```diff
@@ -0,0 +1,37 @@
+    public function up(Schema $schema)
+    {
+        $this->sql('ALTER TABLE transport_domains ADD enabled BOOLEAN NOT NULL');
+        $this->sql('ALTER TABLE payment_domains ADD enabled BOOLEAN NOT NULL');
```
i deleted NOT NULL condition and added queries that will migrate existing domain records to be enabled and all other records in translations will be inserted with enabled set on false.

MattCzerner replied:
Actually i added default value into Entities

PetrHeinz replied:
You would still need to fix the migration - no matter what Doctrine says, these SQL queries will end up with an error...

Default values are not used very much in SSFW, are you sure about this approach?

MattCzerner commented on packages/framework/src/Model/Feed/Standard/StandardFeedItemRepository.php:
```diff
@@ -51,7 +49,6 @@ public function getItems(DomainConfigInterface $domainConfig, $seekItemId, $maxR
             ->addSelect('a')->join('p.calculatedAvailability', 'a')
             ->addSelect('b')->leftJoin('p.brand', 'b')
             ->andWhere('p.variantType != :variantTypeMain')->setParameter('variantTypeMain', Product::VARIANT_TYPE_MAIN)
-            ->join(ProductDomain::class, 'pd', Join::WITH, 'pd.product = p.id AND pd.domainId = :domainId')
```
Well since ProductDomain was not property of Product, there was need to join in query, or i dont understand the question right. But you are right, i added it there.

MattCzerner commented:
So, i think that i fixed everything that you asked me to.

I think that there is only 3 things to discuss. 

First thing is migrations, since i was developing and adding these new ORM things on the run, there was created many migrations. For example Version20180612230916 should be executed earlier then Version20180611123750, because i expect that while im migrating old Transport/Payment Domain data, there is already created unique Constraints, so i dont know what to do with it

Second is, that i added paragraph about domain entities into basics-about-model-architecture.md, so i need to review this

And at last, make these commit messages better, i dont want to do it now, since this CR is already hard so i would love to take care of this with you.

Build is okay, i think that everything is functional.

Except these commit messages, i dont care if you find something and fix it right away, i would just like to see in what i did mistake.

PetrHeinz commented on docs/introduction/basics-about-model-architecture.md:
````diff
@@ -67,15 +67,15 @@ class Product
 ```
 
-### Factory
+## Factory
````
It was indented on purpose - it's a factory for entities, so it should be a child of the *Entity* heading: 
- Basics about the model
- Entity
  - Example
  - Factory
    - Example
- Repository
  - Example
- Facade
  - Example
- Service
  - Example
- Cooperation of layers

It would make sense to either leave it indented, or at least renaming it to *Entity Factory*.

PetrHeinz commented:
I have rebased and squashed your commits, you can see the commits in https://github.com/shopsys/shopsys/compare/dc33388c8...mc-multidomain-original-after-squashing, so you can see any further changes. I will have to rebase it a few times more and it would be a pain to do it in fixup commits...

PetrHeinz commented:
I thoroughly checked the changes, fixed a few things and written proper commit messages and a changelog. Please, check the changes (I would suggest checking out mc-multidomain and soft resetting to mc-multidomain-original-after-squashing to view changes locally).

I do not think we should implement the implicit current domain ID in the scope of this change, but please, create a user story for that.

Send it straight to testing after checking the changes I made :+1:

PetrHeinz commented on project-base/tests/ShopBundle/Database/Model/Category/CategoryDomainTest.php:
```diff
@@ -0,0 +1,119 @@
+    const SECOND_DOMAIN_ID = 2;
+    const DEMONSTRATIVE_SEO_TITLE = 'Demonstrative seo title';
+    const DEMONSTRATIVE_SEO_META_DESCRIPTION = 'Demonstrative seo description';
+    const DEMONSTRATIVE_SEO_H1 = 'Demonstrative seo H1';
```
I would personally avoid using constants for these testing strings, but it's OK to do so.

PetrHeinz commented on project-base/tests/ShopBundle/Database/Model/Category/CategoryDomainTest.php:
```diff
@@ -0,0 +1,119 @@
+        $this->assertTrue($refreshedCategory->isEnabled(self::FIRST_DOMAIN_ID));
+    }
+
+    public function testCreateCategoryNotEnabledOnDomain()
```
not enabled = disabled

MattCzerner replied:
done

PetrHeinz commented on project-base/tests/ShopBundle/Database/Model/Category/CategoryDomainTest.php:
```diff
@@ -0,0 +1,119 @@
+        $categoryData->enabled = [
+            self::FIRST_DOMAIN_ID => true,
+            self::SECOND_DOMAIN_ID => false,
+        ];
```
Here, you are rewriting the whole array, in the other methods you were just rewriting a single item. This inconsistency may have strange consequences when running with different number of domains.

You can achieve what you want without being inconsistent:

```php
$categoryData->enabled[self::FIRST_DOMAIN_ID] = true;
$categoryData->enabled[self::SECOND_DOMAIN_ID] = false;
```

In Payment / Transport tests you even change the number of items - you use single-item arrays.

Not sure if it's a problem that has to be solved...

MattCzerner replied:
I solved it

PetrHeinz commented on project-base/tests/ShopBundle/Database/Model/Product/ProductDomainTest.php:
```diff
@@ -0,0 +1,94 @@
+        parent::setUp();
+        $this->productDataFactory = $this->getContainer()->get(ProductDataFactory::class);
+        $this->productFactory = $this->getContainer()->get(ProductFactory::class);
+        $this->availabilityFacade = $this->getContainer()->get(AvailabilityFacade::class);
```
AvailabilityFacade is never used, persistent reference is used instead.

PetrHeinz commented:
@MattCzerner Please, mention the merge commit (c77083c3c) when you merge PRs manually, otherwise it seem,s like it wasn't merged at all.
